### DISCUSSION TITLE
Is possible custom return Database Realtime onSubscribe & multiple where clause on it ?

### DISCUSSION THREAD

**Discussion by zgramming** [supabase/supabase#3193] [Questions] [answered]
I'm confused whether this should be a separate question or not, but I have 2 questions regarding Realtime Database, and i integrated **Supabase** with **Flutter**

In this [docs](https://supabase.io/docs/reference/javascript/subscribe) about Realtime Database, it is very clear that we can listen for any changes in our table by defining table and action what we want listen (Insert,Update,Delete). And even better, we can listen [to row level change](https://supabase.io/docs/reference/javascript/subscribe#listening-to-row-level-changes).

But i have some special cases which have not been explained in the documentation, let me explain this.

## Table Profile
<table>
<thead>
<tr>
<th>id</th>
<th>id_user</th>
<th>name</th>
<th>email</th>
<th>picture_profile</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>55c0aa7c-d479-41b9-b06a-470c13f35bc1</td>
<td>Zeffry Reynando</td>
<td>zeffry.reynando@gmail.com</td>
<td>https://example.com/image.png</td>
</tr>
<tr>
<td>2</td>
<td>58a3eebf-98c1-46e3-8603-08c266676f43</td>
<td>Icha</td>
<td>icha@gmail.com</td>
<td>https://example.com/image2.png</td>
</tr>
</tbody>
</table>

## Table Inbox
<table>
<thead>
<tr>
<th>id</th>
<th>inbox_channel</th>
<th>id_user</th>
<th>id_sender</th>
<th>message</th>
<th>message_date</th>
</tr>
</thead>
<tbody>
<tr>
<td>1</td>
<td>1_2</td>
<td>1</td>
<td>1</td>
<td>Halo ?</td>
<td>111122223333</td>
</tr>
<tr>
<td>2</td>
<td>1_2</td>
<td>2</td>
<td>1</td>
<td>Halo ?</td>
<td>111122223333</td>
</tr>

</tbody>
</table>

> I have made an example table above, to make it easier to explain what I mean.

# 1. Custom return onSubscribe

```
const mySubscription = supabase
  .from('inbox')
  .on('*', payload => {
    console.log('Change received!', payload)
  })
  .subscribe()
```

When there is a change in the our table (UPDATE,INSERT,DELETE), `onSubscribe` give us return **object**. If we use the example table above, when table **inbox** there is data change `onSubscribe` will give us : 

```
{
 "id": "2",
 "inbox_channel" : "1_2",
 "id_user" : "1",
 "id_sender" : "2",
 "message" : "halooo",
 "message_date" : 1112223333
}
```
Is possible we can custom return callback like this  ? : 

```
 "id": "2",
 "inbox_channel" : "1_2",
 /// It from profile table
 "user" : {
        "id" : "1",
        "id_user" : "55c0aa7c-d479-41b9-b06a-470c13f35bc1",
        "name" : "zeffry",
        "email" : "zeffry.reynando@gmail.com",
        "picture_profile" : "https://example.com/image1.png",
 }
 
 "sender" : {
        "id" : "2",
        "id_user" : "58a3eebf-98c1-46e3-8603-08c266676f43",
        "name" : "icha",
        "email" : "icha@gmail.com",
        "picture_profile" : "https://example.com/image2.png",
 }
 "message" : "halooo",
 "message_date" : 1112223333
}
```
If we use standar select, it's can be achieve using [JOIN](https://supabase.io/docs/reference/javascript/select#query-foreign-tables).

```
supabase.from("inbox")
        .select("*, user:id_user(*), sender:id_sender(*)")
        .eq('id_user', idUser)
        .execute();
```

But i don't know implement it when using `onSubscribe`.

# 2. Multiple Where clause onSubscribe

As I explained above, we can listen to changes down [to row level change](https://supabase.io/docs/reference/javascript/subscribe#listening-to-row-level-changes). In documentation, as far as I understand it is only described using 1 filter. 

```
const mySubscription = supabase
  .from('countries:id=eq.200')
  .on('UPDATE', handleRecordUpdated)
  .subscribe()
```

It's possible use multiple where clause on `onSubscribe` ?  something like this

```
const mySubscription = supabase
  .from('countries:id=eq.200, countries:id_user=eq.100')
  .on('UPDATE', handleRecordUpdated)
  .subscribe()
```

Thank's.

**Comment by dshukertjr** [comment]
Hi @zgramming!

>1. Custom return onSubscribe

This is not possible currently due to the limitation of realtime. What I would do in your case is setup a listener on `inbox` like you have, and also create a variable to store profile data of different users. Whenever you get a new message in the `inbox`, you can search within the profile variable to see if you already have that profile in memory. If not, you can go get it with standard `select()` query. Supabase is really fast, so this should still feel relatively realtime.

Would this be a viable solution in your case?

Also, if you are using Flutter, it might be easier to listen to realtime data using [`stream()`](https://supabase.io/docs/reference/dart/stream) method! Let me know what you think of it!

>2. Multiple Where clause onSubscribe

You would achieve this by creating multiple listeners:

```typescript
const mySubscription = supabase
  .from('countries:id=eq.200')
  .on('UPDATE', handleRecordUpdated)
  .subscribe()
  
  const anotherSubscription = supabase
  .from('countries:id=eq.300')
  .on('UPDATE', handleRecordUpdated)
  .subscribe()
```

**Comment by zgramming** [comment, marked as answer]
Hi @dshukertjr ! thank's for your response. 

> Also, if you are using Flutter, it might be easier to listen to realtime data using stream() method! Let me know what you think of it!

Yes, because my project using flutter and Supabase package support to using [stream](https://pub.dev/packages/supabase/example) in pub.dev, i already implement like this : 

```
final boxProfile = Hive.box<ProfileHiveModel>(Constant.hiveKeyBoxProfile);
final listen = Constant.supabase
      .from("inbox:id_user=eq.${user?.id}")
      .stream()
      .order('inbox_last_message_date')
      .execute()
      .listen((events) async {
    if (events.isNotEmpty) {
      ProfileModel? sender;

      /// Check if sender exists in Hive Box
      /// if exists, we used it from local database
      /// otherwise call API
      for (final event in events) {
        if (boxProfile.containsKey(event['id_sender'])) {
          final getProfile = boxProfile.get(event['id_sender']);
          sender = const ProfileHiveModel().convertToProfileModel(getProfile);
        } else {
          sender = await SupabaseQuery.instance.getUserById(event['id_sender'] as int);
          boxProfile.put(sender.id, const ProfileHiveModel().convertFromProfileModel(sender));
        }

        final inbox = InboxModel.fromJson(event).copyWith(sender: sender);
        ref.read(InboxProvider.provider.notifier).upsertInbox(inbox);
      }
    }
  });
```
A little explain about above code, i listen to table **inbox** where id_user equal to user login id, so i only receive return from stream where have any changes on table **inbox** equal to id_user. Then i check if sender exists in my local database **(Hive in my case)**, if sender exists in the local then i using from local, otherwise i call API to supabase to get user by id. 

It's works what i expected, but i notice return from stream is `List<Map<String,dynamic>>` . The problem about this (IMO), if only single row changes in **Inbox**, it will return all data on **inbox** equal to id_user i have defined before. Let me explain this with table : 

<table>
    <thead>
        <tr>
            <th>id</th>
            <th>inbox_channel</th>
            <th>id_user</th>
            <th>id_sender</th>
            <th>message</th>
            <th>message_date</th>
            <th>X</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>1</td>
            <td>1_2</td>
            <td>1</td>
            <td>2</td>
            <td>Hello Zeffry I icha</td>
            <td>123456</td>
            <td>Only this row have changes on table</td>
        </tr>
        <tr>
            <td>2</td>
            <td>1_3</td>
            <td>1</td>
            <td>3</td>
            <td>Hello Zeffry I riri</td>
            <td>123456</td>
        </tr>
        <tr>
            <td>1</td>
            <td>1_4</td>
            <td>1</td>
            <td>4</td>
            <td>Hello Zeffry I budi</td>
            <td>123456</td>
        </tr>
        <tr>
            <td>1</td>
            <td>1_5</td>
            <td>1</td>
            <td>5</td>
            <td>Hello Zeffry I joko</td>
            <td>123456</td>
        </tr>
    </tbody>
</table>

As I described above, if only **id 1** have changes  in **inbox**, then stream will return all data on **inbox** equal id_user. So stream will return like this : 

## return from stream
```
[{"id":1,"inbox_channel":"1_2","id_user":"1","id_sender":"2","message":"hello zeffry i icha","message_date":"11112222333"},{"id":2,"inbox_channel":"1_3","id_user":"1","id_sender":"3","message":"hello zeffry i riri","message_date":"11112222333"},{"id":1,"inbox_channel":"1_4","id_user":"1","id_sender":"4","message":"hello zeffry i budi","message_date":"11112222333"},{"id":1,"inbox_channel":"1_5","id_user":"1","id_sender":"5","message":"hello zeffry i joko","message_date":"11112222333"}]
```

If we using **onSubscribe**, it will return single data only changes on **inbox** and we have ability to detect is **Delete/Insert/Update** type.

## return from onSubscribe(only data changes return)

```
{"id":1,"inbox_channel":"1_2","id_user":"1","id_sender":"2","message":"hello zeffry i icha","message_date":"11112222333"}
```

The problem i think use **stream** rather than **onSubscribe** is if my **inbox** have large data example 1000 rows with id_user == 1, then if only 1 row changes in **inbox**, stream will return 1000 rows. 
So i ask about best practice to handle this scenario, it's okay if stream will return all data although is only 1 row changes in **inbox** or any recommendation ?


> You would achieve this by creating multiple listeners:

For second question, i think your advice it's not suitable for me. Because you create **2 subscribe**, but i need only **1 subscribe** with multiple where clause. 

If i follow your advice, it will create **2 subscribe** then if subscribe detect changes on table it will multiple return.

So **onSubscribe** can't execute multiple where clause like my example above ? 

```
const mySubscription = supabase
  .from('countries:id=eq.200, countries:id_user=eq.100')
  .on('UPDATE', handleRecordUpdated)
  .subscribe()
```

**Reply by dshukertjr** [reply]
@zgramming Thank you for such a detailed explanation!

>The problem i think use stream rather than onSubscribe is if my inbox have large data example 1000 rows with id_user == 1, then if only 1 row changes in inbox, stream will return 1000 rows.
So i ask about best practice to handle this scenario, it's okay if stream will return all data although is only 1 row changes in inbox or any recommendation ?

1,000 should be fine, but certainly there could be too much data at some point. Have you used `limit()` with stream? You can find the documentation [here](https://supabase.io/docs/reference/dart/stream#with-limit), but with this you can limit the amount of data you load initially from the database. 

You also raise a very good point that with current setup of `stream()`, it's impossible to find out which row was inserted, updated, or deleted. We might need to add this feature in the future releases.

>If i follow your advice, it will create 2 subscribe then if subscribe detect changes on table it will multiple return.

I see, what you want to do is narrow down the result that you get even more with two where clause! Essentially, you want to receive data for a SQL like this:

```sql
select * from countries where id = 200 and id_user = 100;
```

Yeah, this is not possible at the moment unfortunately. There might be more filters coming to realtime in the future as discussed [here](https://github.com/supabase/realtime/issues/132#issuecomment-814593464), but for now, that is the limitation of it.

**Reply by zgramming** [reply]
@dshukertjr not implement stream using `limit` for now, but i think use `limit` it's useless for my case (IMO). simple explain again : 

<table>
    <thead>
        <tr>
            <th>id</th>
            <th>inbox_channel</th>
            <th>id_user</th>
            <th>id_sender</th>
            <th>message</th>
            <th>is_typing</th>
            <th>message_date</th>
            <th>X</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>1</td>
            <td>1_2</td>
            <td>1</td>
            <td>2</td>
            <td>Hello Zeffry I icha</td>
            <td>false</td>
            <td>123456</td>
        </tr>
        <tr>
            <td>2</td>
            <td>1_3</td>
            <td>1</td>
            <td>3</td>
            <td>Hello Zeffry I riri</td>
            <td>false</td>
            <td>123456</td>
        </tr>
        <tr>
            <td>1</td>
            <td>1_4</td>
            <td>1</td>
            <td>4</td>
            <td>Hello Zeffry I budi</td>
            <td>true</td>
            <td>123456</td>
            <td>I Want only listen to this row</td>
        </tr>
        <tr>
            <td>1</td>
            <td>1_5</td>
            <td>1</td>
            <td>5</td>
            <td>Hello Zeffry I joko</td>
            <td>false</td>
            <td>123456</td>
        </tr>
    </tbody>
</table>

If the case like above `i want get status typing in row where id_user = 4 and inbox_channel = 1_4`. For now, like you said it's not possible to use multiple filter in `stream`. 

> Yeah, this is not possible at the moment unfortunately. There might be more filters coming to realtime in the future as discussed here, but for now, that is the limitation of it.

I think use `limit` it's not suitable because i don't know exact location data i want listen. If data at position **4** in database, when i `limit` only **1** then i can't get data i want (it's only example, in real case is more complex than this).

Example is only **5** row, when data grow to **1000** row, it's pain to want listen only **1** row when the `stream` give us **1000** row.


> I see, what you want to do is narrow down the result that you get even more with two where clause! Essentially, you want to receive data for a SQL like this:

Yes, that's exactly what I meant.

For now i think's best option is waiting for future update about more complex query in `stream` or `onSubscribe` 😁

**Reply by dshukertjr** [reply]
>I think use limit it's not suitable because i don't know exact location data i want listen. If data at position 4 in database, when i limit only 1 then i can't get data i want (it's only example, in real case is more complex than this).

>Example is only 5 row, when data grow to 1000 row, it's pain to want listen only 1 row when the stream give us 1000 row.

I see your pain. Just to clarify, `stream()` is just a realtime listener behind the scene, so listening to table using `subscribe()` and `stream()` has no difference in performance, but you are right about `stream()` making it hard to pin point which data has changed. If you have any suggestions of making `stream()` better, let me know! I was thinking of changing the return type of `stream()` to something that the user can extract the changed row every time `stream()` fires. 

>For now i think's best option is waiting for future update about more complex query in stream or onSubscribe 😁

Thanks for understanding! BTW, did you take a look at [walrus](https://github.com/supabase/walrus) repo? It's a new project to update Supabase's realtime engine to support new features like more filters and row level security! I haven't tested it out yet, so I'm not sure if we can do multiple `eq` (where) filters with it, but the product is definitely moving forward pretty fast!

**Reply by zgramming** [reply]
@dshukertjr Interisting repo, i will try to look at that !
Thank's before about your clarification about my case !
